lib/change_stream.js

```diff
@@ -0,0 +1,227 @@
+ * Creates a new Change Stream instance (INTERNAL TYPE, do not instantiate directly)
+ * @class ChangeStream
+ * @param {(Db|Collection)} changeDomain The database or collection against which to create the change stream
+ * @param {Array} pipeline An array of aggregation pipeline stages through which to pass change stream documents
+ * @param {object} [options=null] Optional settings
+ * @param {string} [options.fullDocument=none] Allowed values: ‘none’, ‘lookup’. When set to ‘lookup’, the change stream will include both a delta describing the changes to the document, as well as a copy of the entire document that was changed from some time after the change occurred.
+ * @param {number} [options.maxAwaitTimeMS] The maximum amount of time for the server to wait on new documents to satisfy a change stream query
+ * @param {object} [options.resumeAfter=null] Specifies the logical starting point for the new change stream. This should be the _id field from a previously returned change stream document.
+ * @param {boolean} [options.disableResume=false] Whether the driver should automatically attempt to resume the change stream in the event of a potentially resumable error (such as a network error)
+ * @param {number} [options.batchSize=null] The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}.
+ * @param {object} [options.collation=null] Specify collation settings for operation. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}.
+ * @param {ReadPreference} [options.readPreference=null] The read preference. Defaults to the read preference of the database or collection. See {@link https://docs.mongodb.com/manual/reference/read-preference|read preference documentation}.
+ * @fires ChangeStream#change
+ * @return {ChangeStream} a ChangeStream instance.
+ */
+var ChangeStream = function(changeDomain, pipeline, options) {
+  var self = this;
+  self.changeDomain = changeDomain;
+  self.pipeline = pipeline || [];
+  self.options = options || {};
+
+  // Extract the resume token (if it exists)
+  if (options.resumeAfter) {
+    self.cachedResumeToken = options.resumeAfter;
+  }
+
+  // Extract namespace and serverConfig from changeDomain
+  self.namespace = {};
+  if (changeDomain.collectionName) {
+    self.namespace.database = self.changeDomain.s.db.databaseName;
+    self.namespace.collection = self.changeDomain.collectionName;
+    self.serverConfig = self.changeDomain.s.db.serverConfig;
+  } else if (self.changeDomain.databaseName) {
+    self.namespace.database = self.changeDomain.databaseName;
+    self.serverConfig = self.changeDomain.serverConfig;
+  } else {
+    throw new Error('changeDomain must be either a database or collection.');
+  }
+
+  self.cursor = createChangeStreamCursor(self);
+
+  self.on('newListener', function(eventName) {
+    if (eventName === 'change') {
+      self.changeListenerExists = true;
```

Comment:
I wonder if there is a way we can avoid attaching this `changeListenerExists` property to the change stream?